### PR TITLE
Add `termTypePrec` to `TermLiteral` to allow for `TermLiteral SNat`

### PR DESCRIPTION
**Pull request by martijnbastiaan**
For https://github.com/clash-lang/clash-compiler/pull/2270

## Still TODO:

  - [x] Write a changelog entry (see changelog/README.md)
  - [x] Check copyright notices are up to date in edited files

### CODE REVIEW

File: clash-lib/src/Clash/Core/TermLiteral/TH.hs [hunk 1]
```diff
@@ -25,7 +31,99 @@ dcName' :: DataCon -> String
...
+    constraint typVarName = [t| $(conT termLiteralName) $(varT typVarName) |]
+  constraints <- mapM constraint typeVarNames
+  pure $ [InstanceD Nothing constraints instanceType [termTypePrec, termToData]]
+
+-- | For 'Maybe', constructs:
+--
+-- > termTypePrec n _
+-- >   = let
+-- >       showSpace = showChar ' '
+-- >       precCalls = [termTypePrec 10 (Proxy @a)]
```
Comment (DigitalBrains1) [inline clash-lib/src/Clash/Core/TermLiteral/TH.hs]:
```suggestion
-- >       precCalls = [termTypePrec 11 (Proxy @a)]
```

File: clash-lib/src/Clash/Core/TermLiteral/TH.hs [hunk 2]
```diff
@@ -25,7 +31,99 @@ dcName' :: DataCon -> String
...
+-- >       interspersedPrecCalls = intersperse showSpace precCalls
+-- >       showType = foldl (.) (showString "Maybe") (showSpace : interspersedPrecCalls)
+-- >     in
+-- >       if (n <= 10) then
```
Comment (DigitalBrains1) [inline clash-lib/src/Clash/Core/TermLiteral/TH.hs]:
```suggestion
-- >       if (n > 10) then
```

File: clash-lib/src/Clash/Core/TermLiteral/TH.hs [hunk 3]
```diff
@@ -25,7 +31,99 @@ dcName' :: DataCon -> String
...
+deriveTermTypePrec typName typeVarNames = do
+  TyConI (DataD _ _ typeVars _ _ _) <- reify typName
+  showTypeBody <- mkShowTypeBody typeVars
+  pure (FunD termTypePrecName [Clause [VarP nName, WildP] (NormalB showTypeBody) []])
+ where
+  showTypeName = [| showString $(litE (StringL (nameBase typName))) |]
+
+  -- Constructs:
+  --
+  -- > termTypePrec 10 (Proxy @a)
```
Comment (DigitalBrains1) [inline clash-lib/src/Clash/Core/TermLiteral/TH.hs]:
```suggestion
  -- > termTypePrec 11 (Proxy @a)
```

File: clash-lib/src/Clash/Core/TermLiteral/TH.hs [hunk 4]
```diff
@@ -21,11 +28,114 @@ import           Clash.Core.Name                 (nameOcc)
...
+  --
+  mkTypePrecCall typVarName =
+    [| $(varE termTypePrecName) 11 (Proxy @($(varT typVarName))) |]
+
+  -- Constructs:
+  --
+  -- > showString "Maybe" . showChar ' ' . termTypePrec 11 (Proxy @a)
+  --
+  -- This is wrapped in an if-statement wrapping the result in parentheses if the
+  -- incoming prec is less than 10 (function application).
```
Comment (DigitalBrains1) [inline clash-lib/src/Clash/Core/TermLiteral/TH.hs:115]:
```suggestion
  -- incoming prec is more than 10 (function application).
```

**Review by DigitalBrains1 — COMMENTED** [review]
I think this is going to mention a bunch of things you already fixed... because I started the review on an older version and after pressing "Refresh" it says I have 7 comments but I can only see two.

File: clash-lib/src/Clash/Core/TermLiteral.hs [hunk 5]
```diff
@@ -80,35 +102,55 @@ instance TermLiteral Natural where
...
+    --      replicate that behavior here for ease of testing.
+      showChar '('
+    . termTypePrec 11 (Proxy @a)
+    . showString ","
+    . termTypePrec 11 (Proxy @b)
+    . showChar ')'
+
+$(deriveTermLiteral ''Bool)
+$(deriveTermLiteral ''Maybe)
+$(deriveTermLiteral ''Either)
```
[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply clash-lib/src/Clash/Core/TermLiteral.hs:144]: Yep

**Review by DigitalBrains1 — APPROVED** [review]

File: clash-lib/src/Clash/Core/TermLiteral.hs [hunk 6]
```diff
@@ -47,6 +51,24 @@ class Typeable a => TermLiteral a where
...
+  -- Without it, GHC would ask for a 'KnownNat' constraint on the instance, which
+  -- would defeat the purpose of it.
+  termTypePrec ::
+    -- | The operator precedence of the enclosing context (a number from @0@ to
+    -- @11@). Function application has precedence @10@. Used to determine whether
+    -- the result should be wrapped in parentheses.
+    Int ->
+    -- | Term whose type needs to be pretty printed
+    Proxy a ->
+    -- | Pretty printed type
```
Comment (leonschoorl) [inline clash-lib/src/Clash/Core/TermLiteral.hs:66]: It's [more complicated](https://hackage.haskell.org/package/base-4.17.0.0/docs/Prelude.html#t:ShowS) then that.
Maybe just leave out the description of the result?
I'd say it's not worth it to describe it here.

**Review by leonschoorl — COMMENTED** [review]
Just like [showsPrec](https://hackage.haskell.org/package/base-4.17.0.0/docs/Text-Show.html#v:showsPrec) has [show](https://hackage.haskell.org/package/base-4.17.0.0/docs/Text-Show.html#v:show), I think it makes sense to define a simpler entry function.
So when using it you can do something like:
`showTermType (Proxy @a)`
Instead of having to do:
`termTypePrec 0 (Proxy @a) ""`

Following the naming from "Show", it would make sense to then rename `termTypePrec` to `showsTermTypePrec` (or `showsPrecTermType`?).
I wouldn't put `show` equivalent in the class, like `Show` does, because I don't think it make sense to redefine it for individual instances. it can just be a standalone function.

**Comment by martijnbastiaan** [comment]
Done @leonschoorl

File: clash-lib/src/Clash/Core/TermLiteral.hs [hunk 7]
```diff
@@ -30,13 +33,19 @@ import           GHC.Stack
...
 import qualified Clash.Util.Interpolate          as I
 import qualified Clash.Verification.Internal     as Cv
 
 import           Clash.Core.TermLiteral.TH
 
+-- | Pretty print type 'a'
+showType :: TermLiteral a => Proxy a -> String
+showType proxy = showsTypePrec 0 proxy ""
+
 -- | Tools to deal with literals encoded as a "Term".
```
Comment (DigitalBrains1) [inline clash-lib/src/Clash/Core/TermLiteral.hs:47]:
```suggestion
-- | Tools to deal with literals encoded as a 'Term'.
```

File: clash-lib/src/Clash/Core/TermLiteral.hs [hunk 8]
```diff
@@ -47,6 +56,23 @@ class Typeable a => TermLiteral a where
     -- ^ 'Left' indicates a failure, containing the (sub)term that failed to
     -- translate. 'Right' indicates a success.
 
+  -- | Pretty print the type of a term (for error messages). Its default implementation
+  -- uses 'Typeable' to print the type. Note that this method is there to allow
+  -- an instance for 'SNat' to exist (and other GADTs imposing 'KnownNat').
+  -- Without it, GHC would ask for a 'KnownNat' constraint on the instance, which
+  -- would defeat the purpose of it.
```
Comment (DigitalBrains1) [inline clash-lib/src/Clash/Core/TermLiteral.hs:63]:
```suggestion
  -- an instance for 'SNat' to exist (and other GADTs imposing
  -- t'GHC.TypeNats.KnownNat'). Without it, GHC would ask for a @KnownNat@
  -- constraint on the instance, which would defeat the purpose of it.
```

It's not in scope.

File: clash-lib/src/Clash/Core/TermLiteral.hs [hunk 9]
```diff
@@ -30,13 +33,19 @@ import           GHC.Stack
...
 import           Clash.Core.Literal
 import           Clash.Core.Pretty               (showPpr)
+import           Clash.Promoted.Nat
+import           Clash.Promoted.Nat.Unsafe
 import qualified Clash.Util.Interpolate          as I
 import qualified Clash.Verification.Internal     as Cv
 
 import           Clash.Core.TermLiteral.TH
 
+-- | Pretty print type 'a'
```
Comment (DigitalBrains1) [inline clash-lib/src/Clash/Core/TermLiteral.hs:43]:
```suggestion
-- | Pretty print type @a@
```

File: clash-lib/src/Clash/Core/TermLiteral/TH.hs [hunk 10]
```diff
@@ -21,11 +28,120 @@ import           Clash.Core.Name                 (nameOcc)
...
+#if __GLASGOW_HASKELL__ >= 900
+  PlainTV typVarName () -> pure typVarName
+  KindedTV typVarName () StarT -> pure typVarName
+#else
+  PlainTV typVarName -> pure typVarName
+  KindedTV typVarName StarT -> pure typVarName
+#endif
+  k@(KindedTV {}) -> fail $ "Not supported: KindedTV: " <> show k
+
+-- | Derive a 'TermLiteral' instance for given type.
```
Comment (DigitalBrains1) [inline clash-lib/src/Clash/Core/TermLiteral/TH.hs:68]:
```suggestion
-- | Derive a t'Clash.Core.TermLiteral.TermLiteral' instance for given type
```

It's not in scope. Also, it's a sentence fragment if I'm editing it anyway :-)

**Review by DigitalBrains1 — COMMENTED** [review]
I noticed some Haddock thingies.

File: clash-lib/src/Clash/Core/TermLiteral.hs (hunk 8 above)
Comment (martijnbastiaan) [reply clash-lib/src/Clash/Core/TermLiteral.hs:63]: Did you add the `t` on purpose?

File: clash-lib/src/Clash/Core/TermLiteral/TH.hs (hunk 10 above)
Comment (martijnbastiaan) [reply clash-lib/src/Clash/Core/TermLiteral/TH.hs]: Did you add the `t` on purpose?

File: clash-lib/src/Clash/Core/TermLiteral.hs [hunk 11]
```diff
@@ -80,34 +106,60 @@ instance TermLiteral Natural where
...
+    Right (unsafeSNat n)
+  termToData t = Left t
+
+  showsTypePrec n _
+    -- We don't know the literal /n/ at this point. However, we can't simply put
+    -- and /n/ here either, as it might collide with other type variables. To
+    -- prevent confusion, we put an underscore. This is obviously "wrong", but
+    -- good enough for error messages - the main purpose of this function.
+    | n > 10    = showString "(SNat _)"
+    | otherwise = showString "SNat _"
```
Comment (DigitalBrains1) [inline clash-lib/src/Clash/Core/TermLiteral.hs:126]:
```suggestion
    = showParen (n > 10) $ showString "SNat _"
```

File: clash-lib/src/Clash/Core/TermLiteral/TH.hs [hunk 12]
```diff
@@ -21,11 +28,120 @@ import           Clash.Core.Name                 (nameOcc)
...
+-- >   = let
+-- >       showSpace = showChar ' '
+-- >       precCalls = [showsTypePrec 11 (Proxy @a)]
+-- >       interspersedPrecCalls = intersperse showSpace precCalls
+-- >       showType = foldl (.) (showString "Maybe") (showSpace : interspersedPrecCalls)
+-- >     in
+-- >       if (n > 10) then
+-- >         showChar '(' . showType . showChar ')'
+-- >       else
+-- >         showType
```
Comment (DigitalBrains1) [inline clash-lib/src/Clash/Core/TermLiteral/TH.hs:95]:
```suggestion
-- >       showParen (n > 10) showType
```

File: clash-lib/src/Clash/Core/TermLiteral/TH.hs [hunk 13]
```diff
@@ -21,11 +28,120 @@ import           Clash.Core.Name                 (nameOcc)
...
+        let
+          showSpace = showChar ' '
+          precCalls = $(listE (map mkTypePrecCall typeVarNames))
+          interspersedPrecCalls = intersperse showSpace precCalls
+          showType = foldl (.) $(showTypeName) (showSpace : interspersedPrecCalls)
+        in
+          if $(varE nName) > 10 then
+            showChar '(' . showType . showChar ')'
+          else
+            showType
```
Comment (DigitalBrains1) [inline clash-lib/src/Clash/Core/TermLiteral/TH.hs:141]:
```suggestion
          showParen ($(varE nName) > 10) showType
```

**Review by DigitalBrains1 — COMMENTED** [review]
There's a convenience function for parentheses that makes the code short and sweet

File: clash-lib/src/Clash/Core/TermLiteral/TH.hs (hunk 10 above)
Comment (DigitalBrains1) [reply clash-lib/src/Clash/Core/TermLiteral/TH.hs]: Answered [here](https://github.com/clash-lang/clash-compiler/pull/2329#discussion_r971732826)

**Pull request merged**